crates/extensions/c8y_mapper_ext/src/operations/mod.rs

```diff
@@ -13,47 +13,270 @@
+use tedge_actors::LoggingSender;
+use tedge_actors::Sender;
 use tedge_api::commands::ConfigMetadata;
+use tedge_api::entity_store::EntityExternalId;
+use tedge_api::mqtt_topics::Channel;
 use tedge_api::mqtt_topics::EntityTopicId;
-use tedge_api::workflow::GenericCommandState;
+use tedge_api::mqtt_topics::IdGenerator;
+use tedge_api::mqtt_topics::MqttSchema;
+use tedge_api::mqtt_topics::OperationType;
 use tedge_api::Jsonify;
+use tedge_config::AutoLogUpload;
+use tedge_config::SoftwareManagementApiFlag;
 use tedge_mqtt_ext::MqttMessage;
+use tedge_mqtt_ext::Topic;
 use tracing::error;
 
 pub mod config_snapshot;
 pub mod config_update;
 pub mod firmware_update;
 pub mod log_upload;
+mod restart;
+mod software_list;
+mod software_update;
+mod upload;
 
-/// Represents a pending download performed by the downloader from the FTS.
+/// Handles operations.
 ///
-/// Functions which download files from the tedge File Transfer Service as part of handling
-/// operations (e.g. when performing `log_upload` or `config_snapshot`, the relevant file is
-/// uploaded into FTS) will use this type for communicating with the Downloader actor.
-pub struct FtsDownloadOperationData {
-    pub download_type: FtsDownloadOperationType,
-    pub url: String,
+/// Handling an operation usually consists of 3 steps:
+///
+/// 1. Receive a smartrest message which is an operation request, convert it to thin-edge message,
+///    and publish on local MQTT (done by the converter).
+/// 2. Various local thin-edge components/services (e.g. tedge-agent) execute the operation, and
+///    when they're done, they publish an MQTT message with 'status: successful/failed'
+/// 3. The cumulocity mapper needs to do some additional steps, like downloading/uploading files via
+///    HTTP, or talking to C8y via HTTP proxy, before it can send operation response via the bridge
+///    and then clear the local MQTT operation topic.
+///
+/// This struct concerns itself with performing step 3.
+///
+/// Incoming operation-related MQTT messages need to be passed to the [`Self::handle`] method, which
+/// performs an operation in the background in separate tasks. The operation tasks themselves handle
+/// reporting their success/failure.
+pub struct OperationHandler {
+    pub capabilities: Capabilities,
+    pub auto_log_upload: AutoLogUpload,
+    pub tedge_http_host: Arc<str>,
+    pub tmp_dir: Arc<Utf8Path>,
+    pub mqtt_schema: MqttSchema,
+    pub c8y_prefix: tedge_config::TopicPrefix,
+    pub software_management_api: SoftwareManagementApiFlag,
+    pub command_id: IdGenerator,
+
+    pub http_proxy: C8YHttpProxy,
+    pub c8y_endpoint: C8yEndPoint,
+    pub auth_proxy: ProxyUrlGenerator,
+
+    pub downloader: ClientMessageBox<IdDownloadRequest, IdDownloadResult>,
+    pub uploader: ClientMessageBox<IdUploadRequest, IdUploadResult>,
+    pub mqtt_publisher: LoggingSender<MqttMessage>,
+
+    pub running_operations: Arc<Mutex<HashMap<Arc<str>, RunningOperation>>>,
```

Comment:
Why was this map required at this `OperationHandler` level? The `RunningOperation` has no use of it. From what I understand, this was only required at the converter level for book-keeping the spawned tasks and the message routing. I feel that even the message routing to these tasks could be moved into the converter level so that we can avoid mutable state here and avoid concurrent accesses to it as well.

Reply:
> Why was this map required at this OperationHandler level? The RunningOperation has no use of it.

`RunningOperation` doesn't use it, but it uses other fields of `OperationHandler`. They should ideally be in a separate object, but passing all the properties from `CumulocityConverter` is a pain. Before finalizing the PR, I'll try one approach I had in mind to simplify this.

> From what I understand, this was only required at the converter level for book-keeping the spawned tasks and the message routing.

Yes, the idea is that we have to route messages received via MQTT into the task that handle an operation with a given id. But I don't understand what you mean when you say "it was required at the converter level". Do you mean it was required in the converter before my PR, or do you mean something else?

> I feel that even the message routing to these tasks could be moved into the converter level so that we can avoid mutable concurrent accesses to it as well.

I wanted to hide from the converter how exactly the messages get passed into tasks, and only expose a simple interface - the converter only calls `handle()` and the rest of operation handling happens inside this module. So I'm not really sure what you mean by "move routing into the converter level" and what other levels we're talking about.

Reply:
> RunningOperation doesn't use it, but it uses other fields of OperationHandler. They should ideally be in a separate object, but passing all the properties from CumulocityConverter is a pain. Before finalizing the PR, I'll try one approach I had in mind to simplify this.

The handler was cleaned up a lot in c4e27504de95d47aa8a3ac598eeec2397e4d74b5, mutex on `running_operations` was removed.

Reply:
> I wanted to hide from the converter how exactly the messages get passed into tasks, and only expose a simple interface - the converter only calls handle() and the rest of operation handling happens inside this module. So I'm not really sure what you mean by "move routing into the converter level" and what other levels we're talking about.

Based on that description of how you wanted to abstract the `OperationHandler` without overloading the converter, I see 3 different levels here:
1. Converter: Just converts messages and passes it to operation handler
2. Operation handler: Maintains the map of running operations and routes messages to them
3. The configs and handles (the proxy, message boxes, senders etc) required by the `RunningOperation` to handle individual operations

I think you've combined both 2 and 3 into that single `OperationHandler` struct which is causing this confusion. If they were separated out, then the container can just have a single `OperationHandler` instance in it, with no `Arc<Mutex>>` guard required over it and the `RunningOperation` instances can just use clones of the rest of that struct, without no shared mutable access required between them either.

> RunningOperation doesn't use it, but it uses other fields of `OperationHandler`. They should ideally be in a separate object, but passing all the properties from CumulocityConverter is a pain. Before finalizing the PR, I'll try one approach I had in mind to simplify this.

Yes, I think that's the way to go.

> But I don't understand what you mean when you say "it was required at the converter level". Do you mean it was required in the converter before my PR, or do you mean something else?

I was proposing to move that `running_operations` maintenance responsibility into the converter, but your proposal to keep those with the `OperationHandler` struct sounds better, as long as the rest of the shared fields are moved out of it.

Reply:
> The handler was cleaned up a lot in https://github.com/thin-edge/thin-edge.io/commit/c4e27504de95d47aa8a3ac598eeec2397e4d74b5, mutex on running_operations was removed.

Ignore my previous comment. I guess I was replying to an older stale view of your PR, and I missed this comment. From a quick look, extracting the `OperationContext` abstraction from `OperationHandler` was exactly what was required and that precisely addresses my concerns.